packages/cli/src/s3-store.ts

```diff
@@ -1,181 +1,176 @@
-  }
-
-  get networkName(): string {
-    return this.#storeMap.networkName
+    this.#cache = new Map()
+    this.#logger = logger
+  }
+
+  storeRoot(networkName = this.#networkName) {
+    return `${this.#bucketName}/ceramic/${networkName}`
+  }
+
+  async open(name?: string): Promise<S3KVStore> {
+    const fromCache = this.#cache.get(name)
+    if (fromCache) return fromCache
+    return this.create(name)
+  }
+
+  create(name: string | undefined): Promise<S3KVStore> {
+    return this.#createMutex.use(async () => {
+      const fromCache = this.#cache.get(name)
+      if (fromCache) return fromCache
+      const storePath = await this.dirPath(name)
+      const levelDown = this.#customEndpoint
+        ? new S3LevelDOWN(
+            storePath,
+            new AWSSDK.S3({
+              endpoint: this.#customEndpoint,
+            })
+          )
+        : new S3LevelDOWN(storePath)
+      const levelUp = new LevelUp(levelDown)
+      const kv = new S3KVStore(levelUp)
+      this.#cache.set(name, kv)
+      return kv
+    })
   }
 
-  async init(): Promise<void> {
-    // Check if ELP bucket is used
-    if (this.networkName === Networks.MAINNET) {
+  async dirPath(name: string | undefined) {
+    let storeRoot = this.storeRoot()
+    // Use "elp" store root if on mainnet and the folder is present
+    if (this.#networkName === Networks.MAINNET) {
       const s3 = new AWSSDK.S3()
       const res = await s3
         .listObjectsV2({ Bucket: this.#bucketName, Prefix: 'ceramic/elp', MaxKeys: 1 })
         .promise()
       if (res.Contents?.length) {
         // state store exists and needs to be used
-        this.#diagnosticsLogger.warn(
+        this.#logger.warn(
           `S3 bucket found with ELP location, using it instead of default mainnet location for state store`
         )
-        // Re-create store map with 'elp' network name
-        this.#storeMap = new S3StoreMap('elp' as Networks, this.#bucketName, this.#customEndpoint)
+        storeRoot = this.storeRoot('elp')
       }
     }
+
+    if (name) {
+      return `${storeRoot}/${this.#defaultLocation}-${name}`
+    } else {
+      return `${storeRoot}/${this.#defaultLocation}`
+    }
   }
 
-  async close(useCaseName?: string): Promise<void> {
-    const store = await this.#storeMap.get(useCaseName)
-    await store.close()
+  async close(): Promise<void> {
+    for (const store of this.#cache.values()) {
+      await store.close()
+    }
   }
+}
 
-  async isEmpty(params?: StoreSearchParams): Promise<boolean> {
+class S3KVStore implements IKVStore {
+  readonly level: LevelUp.LevelUp
+
+  readonly #loadingLimit: PQueue
+
+  constructor(level: LevelUp.LevelUp) {
+    this.level = level
+    this.#loadingLimit = new PQueue({
+      intervalCap: MAX_LOAD_RPS,
+      interval: 1000,
+      carryoverConcurrencyCount: true,
+    })
+  }
+
+  init(): Promise<void> {
+    throw new Error('Method not implemented. And should not be used TBH') // TODO
```

Comment:
From the error message it says it shouldn not be used, what is the `TODO` for here?

Reply:
Removed!